test/fake_code_signer/pe_wrapper.rb

```diff
@@ -0,0 +1,123 @@
+class FakeCodeSigner
+  # This class exists to navigate and manipulate the data of the Windows
+  # executable format, PE, see: https://en.wikipedia.org/wiki/Portable_Executable
+  # also https://msdn.microsoft.com/en-us/library/ms809762.aspx
+  class PEWrapper
+    # size of a DWORD is 2 words (4 bytes)
+    DWORD_SIZE = 4
+
+    # struct IMAGE_DOS_HEADER
+    # https://www.nirsoft.net/kernel_struct/vista/IMAGE_DOS_HEADER.html
+    DOS_HEADER_SIZE = 64
+
+    # first field from IMAGE_NT_HEADERS
+    # https://msdn.microsoft.com/en-us/library/windows/desktop/ms680336(v=vs.85).aspx
+    PE_SIGNATURE_SIZE = 4
+
+    # struct IMAGE_FILE_HEADER
+    # see: https://msdn.microsoft.com/en-us/library/windows/desktop/ms680313(v=vs.85).aspx
+    IMAGE_FILE_HEADER_SIZE = 20
+
+    # struct IMAGE_OPTIONAL_HEADER
+    # https://msdn.microsoft.com/en-us/library/windows/desktop/ms680339(v=vs.85).aspx
+    IMAGE_OPTIONAL_HEADER_SIZE = 224
+
+    # http://bytepointer.com/resources/pietrek_in_depth_look_into_pe_format_pt1_figures.htm
+    NUMBER_OF_DATA_DIRECTORY_ENTRIES = 16
+
+    # struct IMAGE_DATA_DIRECTORY
+    # https://msdn.microsoft.com/en-us/library/windows/desktop/ms680305(v=vs.85).aspx
+    DATA_DIRECTORY_ENTRY_SIZE = 8
+
+    def initialize(image)
+      @image = image
+    end
+
+    # set digital signature address in security header
+    def security_address=(offset)
+      @image[security_address_offset, DWORD_SIZE] = raw_bytes(offset)
+    end
+
+    # set digital signature size in security header
+    def security_size=(size)
+      @image[security_size_offset, DWORD_SIZE] = raw_bytes(size)
+    end
+
+    # location of the digital signature
+    def security_address
+      deref(security_address_offset)
+    end
+
+    # size of the digital signature
+    def security_size
+      deref(security_size_offset)
+    end
+
+    # append data to end of executable
+    def append_data(byte_string)
+      @image << byte_string
+    end
+
+    # string representation of this object
+    def to_s
+      @image
+    end
+
+    private
+
+    # convert an integer to a raw byte string
+    def raw_bytes(int)
```

Comment:
`to_unsigned_long`?

Reply:
@jakedouglas hm i can see your point, but it's still the raw byte representation of the `ulong` that we want. I guess a more accurate name would be `raw_bytes_for_unsigned_long` lol. 

I think that `to_unsigned_long` would be a better name for the `deref` method as it takes the raw byte string and converts it to an unsigned long (or whatever tf it is in Ruby), but i actually like `deref` there too as it closely matches the analogous C code in which it's an explicit pointer dereference 😅

Reply:
Sure, it was mostly the distinction that it only converts to one type that I thought would be nice to have in the method name. Makes it more obvious what's going on at the call site.